File: app/src/main/java/android1601/itstep/org/kidsgame/program/data/Gifts.java
package android1601.itstep.org.kidsgame.program.data;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;

import com.franmontiel.localechanger.LocaleChanger;
import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.structure.BaseModel;

import android1601.itstep.org.kidsgame.R;
import android1601.itstep.org.kidsgame.program.KidsKotlinApplication;
import android1601.itstep.org.kidsgame.program.Utility.Utility;
import android1601.itstep.org.kidsgame.program.db_utility.AppDatabase;
import android1601.itstep.org.kidsgame.program.enums.POSITION_TYPE;
import android1601.itstep.org.kidsgame.program.enums.SECTIONS_TYPE;

/**
 * Created by roman on 13.03.2017.
 */

@Table(database = AppDatabase.class)
public class Gifts extends BaseModel implements Parcelable {

    @PrimaryKey(autoincrement = true)
    private long id;

    @Column
    private boolean unlock;
    @Column
    private String resName;

    @Column
    private int positionType;

    @Column
    private int objectSection_id;

    private int imageResId;
    private int silhouetteResId;
    private int nameResId;
    private int voiceResId;
    private int soundResId;

    protected Gifts(Parcel in) {
        id = in.readLong();
        unlock = in.readByte() != 0;
        resName = in.readString();
        positionType = in.readInt();
        objectSection_id = in.readInt();
    }

    public static final Creator<Gifts> CREATOR = new Creator<Gifts>() {
        @Override
        public Gifts createFromParcel(Parcel in) {
            return new Gifts(in);
        }

        @Override
        public Gifts[] newArray(int size) {
            return new Gifts[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(id);
        dest.writeByte((byte) (unlock ? 1 : 0));
        dest.writeString(resName);
        dest.writeInt(positionType);
        dest.writeInt(objectSection_id);
    }

    public Gifts() {
    }

    public long getId() {
        return id;
    }

    public boolean isUnlock() {
        return unlock;
    }

    public String getResName() {
        return resName;
    }

    public int getObjectSection_id() {
        return objectSection_id;
    }

    public int getPositionType() {
        return positionType;
    }

    public POSITION_TYPE getPositionEnum(){
        return POSITION_TYPE.fromPositions(positionType);
    }

    public SECTIONS_TYPE getSectionType(){
        return SECTIONS_TYPE.sections_type(objectSection_id);
    }

    public void setObjectSection_id(int objectSection_id) {
        this.objectSection_id = objectSection_id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public void setUnlock(boolean unlock) {
        this.unlock = unlock;
    }

    public void setResName(String resName) {
        this.resName = resName;
    }



    public int getImageResId(){
        return Utility.getDrawableResourceIdByName(resName);
    }
    public String getSilhouetteResName(){
        return resName+"_inactive";
    }

    public int getSilhouetteResId(){
        return Utility.getDrawableResourceIdByName(getSilhouetteResName());
    }

    public int getTextId(){
        if (nameResId==0)
            nameResId = resName != null? Utility.getStringResourceIdByName(resName) : 0;
        return nameResId;
    }


    public String getVoiceResName(){
        return resName+"_voice"+R.string.sound_locale;
    }

/*    public int getVoiceRawId(){

        return Utility.getRawIdByName(getVoiceResName());
    }*/

    public String getSoundResName(){
        return resName+"_sound";
    }

    public int getVoiceRawId(){
        Context context = KidsKotlinApplication.Companion.getInstance();
        String locale = LocaleChanger.getLocale().getLanguage();
        if (voiceResId == 0)
            voiceResId = resName!=null?Utility.getRawIdByName(resName+"_"+locale):0;
        return voiceResId;
    }

    public int getSoundRawId(){
        Context context = KidsKotlinApplication.Companion.getInstance();
        if (soundResId == 0)
            soundResId = resName!=null?Utility.getRawIdByName(resName+"_sound"):0;
        return soundResId;
    }

    public void setPositionType(int positionType) {
        this.positionType = positionType;
    }

}

File: app/src/main/java/android1601/itstep/org/kidsgame/program/data/Gifts.kt
package android1601.itstep.org.kidsgame.program.data

import android.os.Parcel
import android.os.Parcelable
import android1601.itstep.org.kidsgame.R
import android1601.itstep.org.kidsgame.program.KidsApplication
import android1601.itstep.org.kidsgame.program.Utility.Utility
import android1601.itstep.org.kidsgame.program.db_utility.AppDatabase
import android1601.itstep.org.kidsgame.program.enums.POSITION_TYPE
import android1601.itstep.org.kidsgame.program.enums.SECTIONS_TYPE
import com.raizlabs.android.dbflow.annotation.Column
import com.raizlabs.android.dbflow.annotation.PrimaryKey
import com.raizlabs.android.dbflow.annotation.Table
import com.raizlabs.android.dbflow.structure.BaseModel

/**
 * Created by roman on 13.03.2017.
 */


@Table(database = AppDatabase::class)
class Gifts constructor() : BaseModel(), Parcelable {

    @PrimaryKey(autoincrement = true)
    var id: Long = 0

    @Column
    var isUnlock: Boolean = false
    @Column
    var resName: String? = null

    @Column
    var positionType: Int = 0

    @Column
    var objectSectionId: Int = 0

    val imageResId = Utility.getDrawableResourceIdByName(resName)

    private var nameResId: Int = 0
    private var voiceResId: Int = 0
    private var soundResId: Int = 0

    val positionEnum: POSITION_TYPE
        get() = POSITION_TYPE.fromPositions(positionType)

    val sectionType: SECTIONS_TYPE
        get() = SECTIONS_TYPE.sections_type(objectSectionId)

    val silhouetteResName: String
        get() = resName!! + "_inactive"

    val silhouetteResId = Utility.getDrawableResourceIdByName(silhouetteResName)

    val textId: Int
        get() {
            if (nameResId == 0)
                nameResId = if (resName != null) Utility.getStringResourceIdByName(resName!!) else 0
            return nameResId
        }


    val voiceResName: String
        get() = resName + "_voice" + R.string.sound_locale

    /*    public int getVoiceRawId(){

        return Utility.getRawIdByName(getVoiceResName());
    }*/

    val soundResName: String
        get() = resName!! + "_sound"

    val voiceRawId: Int
        get() {
            val context = KidsApplication.instance
            if (voiceResId == 0)
                voiceResId = if (resName != null) Utility.getRawIdByName(resName + "_" + context!!.getString(R.string.sound_locale)) else 0
            return voiceResId
        }

    val soundRawId: Int
        get() {
            val context = KidsApplication.instance
            if (soundResId == 0)
                soundResId = if (resName != null) Utility.getRawIdByName(resName!! + "_sound") else 0
            return soundResId
        }

    constructor(`in`: Parcel) : this() {
        id = `in`.readLong()
        isUnlock = `in`.readByte().toInt() != 0
        resName = `in`.readString()
        positionType = `in`.readInt()
        objectSectionId = `in`.readInt()

    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(id)
        dest.writeByte((if (isUnlock) 1 else 0).toByte())
        dest.writeString(resName)
        dest.writeInt(positionType)
        dest.writeInt(objectSectionId)
    }


    companion object CREATOR : Parcelable.Creator<Gifts> {
        override fun createFromParcel(parcel: Parcel): Gifts {
            return Gifts(parcel)
        }

        override fun newArray(size: Int): Array<Gifts?> {
            return arrayOfNulls(size)
        }
    }

}
